File: work/GhidraBatchDecompile.java
// Ghidra headless batch decompiler for Black & White static recompilation
// Run via: analyzeHeadless <project_dir> <project_name> -process <binary>
//          -noanalysis -scriptPath <this_dir>
//          -postScript GhidraBatchDecompile.java <addresses.txt> <output.txt>
//
// @author sp00nz/Claude
// @category Decompilation
// @description Batch decompile functions at specified addresses to C pseudocode

import ghidra.app.decompiler.*;
import ghidra.app.script.GhidraScript;
import ghidra.program.model.address.*;
import ghidra.program.model.listing.*;
import ghidra.util.task.ConsoleTaskMonitor;
import java.io.*;
import java.util.*;

public class GhidraBatchDecompile extends GhidraScript {

    @Override
    public void run() throws Exception {
        String[] args = getScriptArgs();

        if (args.length < 1) {
            println("Usage: GhidraBatchDecompile.java <addresses.txt> [output.txt]");
            return;
        }

        String addrFile = args[0];
        String outFile = args.length > 1 ? args[1] :
            addrFile.replace(".txt", "_decompiled.txt");

        println("=== Ghidra Batch Decompiler ===");
        println("Binary: " + currentProgram.getName());
        println("Address file: " + addrFile);
        println("Output file: " + outFile);

        // Read addresses
        List<Long> addresses = readAddresses(addrFile);
        println("Loaded " + addresses.size() + " addresses");

        // Set up decompiler
        DecompInterface decomp = new DecompInterface();
        DecompileOptions opts = new DecompileOptions();
        decomp.setOptions(opts);
        decomp.openProgram(currentProgram);

        long startTime = System.currentTimeMillis();
        int success = 0, failed = 0, skipped = 0;

        try (PrintWriter out = new PrintWriter(new FileWriter(outFile))) {
            out.println("// ============================================================");
            out.println("// Black & White Batch Decompilation Output");
            out.println("// Binary: " + currentProgram.getName());
            out.println("// Addresses: " + addresses.size());
            out.println("// Date: " + new Date());
            out.println("// ============================================================");
            out.println();

            for (int i = 0; i < addresses.size(); i++) {
                long addrLong = addresses.get(i);
                String addrHex = String.format("0x%08x", addrLong);

                Address addr = currentProgram.getAddressFactory()
                    .getDefaultAddressSpace().getAddress(addrLong);

                // Try exact match first, then containing function
                Function func = currentProgram.getFunctionManager().getFunctionAt(addr);
                if (func == null) {
                    func = getFunctionContaining(addr);
                }
                // If still no function, try creating one at this address
                if (func == null) {
                    try {
                        ghidra.app.cmd.function.CreateFunctionCmd cmd =
                            new ghidra.app.cmd.function.CreateFunctionCmd(addr);
                        cmd.applyTo(currentProgram, monitor);
                        func = currentProgram.getFunctionManager().getFunctionAt(addr);
                    } catch (Exception e2) {
                        // Ignore creation failure
                    }
                }
                if (func == null) {
                    println("[" + (i+1) + "/" + addresses.size() + "] SKIP: " + addrHex + " — no function");
                    out.println("// SKIP: " + addrHex + " — no function at this address");
                    out.println();
                    skipped++;
                    continue;
                }

                String funcName = func.getName();
                long funcSize = func.getBody().getNumAddresses();
                String entryHex = func.getEntryPoint().toString();
                boolean exactMatch = func.getEntryPoint().equals(addr);

                try {
                    DecompileResults results = decomp.decompileFunction(func, 60, monitor);

                    if (results.decompileCompleted()) {
                        String cCode = results.getDecompiledFunction().getC();

                        out.println("// ========================================");
                        out.println("// [" + (i+1) + "] " + funcName);
                        out.println("// Requested: " + addrHex);
                        if (!exactMatch) {
                            out.println("// Entry point: " + entryHex + " (containing function)");
                        }
                        out.println("// Size: " + funcSize + " bytes");
                        out.println("// ========================================");
                        out.println();
                        out.println(cCode);
                        out.println();

                        success++;
                        println("[" + (i+1) + "/" + addresses.size() + "] OK: " + funcName +
                                " @ " + addrHex + " (" + funcSize + " bytes)");
                    } else {
                        String errMsg = results.getErrorMessage();
                        if (errMsg == null) errMsg = "unknown error";
                        out.println("// FAILED: " + funcName + " @ " + addrHex + " — " + errMsg);
                        out.println();
                        failed++;
                        println("[" + (i+1) + "/" + addresses.size() + "] FAILED: " + funcName);
                    }
                } catch (Exception e) {
                    out.println("// ERROR: " + funcName + " @ " + addrHex + " — " + e.getMessage());
                    out.println();
                    failed++;
                    println("[" + (i+1) + "/" + addresses.size() + "] ERROR: " + funcName + " — " + e.getMessage());
                }
            }

            long elapsed = System.currentTimeMillis() - startTime;
            double rate = success * 1000.0 / Math.max(elapsed, 1);

            out.println("// ============================================================");
            out.println("// SUMMARY");
            out.println("// Total: " + addresses.size() +
                        " | Success: " + success +
                        " | Failed: " + failed +
                        " | Skipped: " + skipped);
            out.println("// Time: " + (elapsed / 1000) + " seconds (" +
                        String.format("%.2f", rate) + " func/sec)");
            out.println("// ============================================================");
        } finally {
            decomp.closeProgram();
        }

        long elapsed = System.currentTimeMillis() - startTime;
        println("\n=== Complete ===");
        println("Success: " + success + " / " + addresses.size());
        println("Failed: " + failed + ", Skipped: " + skipped);
        println("Time: " + (elapsed / 1000) + " seconds");
        println("Output: " + outFile);
    }

    private List<Long> readAddresses(String filepath) throws IOException {
        List<Long> addresses = new ArrayList<>();
        try (BufferedReader reader = new BufferedReader(new FileReader(filepath))) {
            String line;
            while ((line = reader.readLine()) != null) {
                line = line.trim();
                if (line.isEmpty() || line.startsWith("//") || line.startsWith("#")) continue;

                // Strip inline comments
                int commentIdx = line.indexOf("//");
                if (commentIdx > 0) line = line.substring(0, commentIdx).trim();
                commentIdx = line.indexOf("#");
                if (commentIdx > 0) line = line.substring(0, commentIdx).trim();

                try {
                    String hexStr = line.startsWith("0x") || line.startsWith("0X") ?
                        line.substring(2) : line;
                    addresses.add(Long.parseLong(hexStr, 16));
                } catch (NumberFormatException e) {
                    // Skip invalid lines
                }
            }
        }
        return addresses;
    }
}

File: work/ghidra_batch_decompile.py
# Ghidra headless batch decompiler for Black & White static recompilation
# Run via: analyzeHeadless <project_dir> <project_name> -import <binary> \
#          -scriptPath <this_dir> -postScript ghidra_batch_decompile.py <addresses.txt> <output.txt>
#
# Or for an already-imported project:
#   analyzeHeadless <project_dir> <project_name> -noanalysis \
#   -scriptPath <this_dir> -postScript ghidra_batch_decompile.py <addresses.txt> <output.txt>
#
# @author sp00nz/Claude
# @category Decompilation
# @description Batch decompile functions at specified addresses

from ghidra.app.decompiler import DecompInterface, DecompileOptions
from ghidra.util.task import ConsoleTaskMonitor
import os
import sys
import time

def read_addresses(filepath):
    """Read hex addresses from file, one per line. Skips comments and blanks."""
    addresses = []
    with open(filepath, 'r') as f:
        for line in f:
            line = line.strip()
            if not line or line.startswith('//') or line.startswith('#'):
                continue
            # Strip inline comments
            if '//' in line:
                line = line[:line.index('//')].strip()
            if '#' in line:
                line = line[:line.index('#')].strip()
            try:
                addr = int(line, 16) if line.startswith('0x') or line.startswith('0X') else int(line, 16)
                addresses.append(addr)
            except ValueError:
                print("WARNING: Skipping invalid address: " + line)
    return addresses

def get_function_at(program, addr_long):
    """Get or create a function at the given address."""
    addr_factory = program.getAddressFactory()
    addr = addr_factory.getDefaultAddressSpace().getAddress(addr_long)

    # Try to get existing function
    func = getFunctionContaining(addr)
    if func and func.getEntryPoint().equals(addr):
        return func

    # Try exact match
    func = program.getFunctionManager().getFunctionAt(addr)
    if func:
        return func

    # Function might not be defined — try to create it
    from ghidra.app.cmd.function import CreateFunctionCmd
    cmd = CreateFunctionCmd(addr)
    cmd.applyTo(program, ConsoleTaskMonitor())
    return program.getFunctionManager().getFunctionAt(addr)

def run():
    args = getScriptArgs()

    if len(args) < 1:
        print("Usage: ghidra_batch_decompile.py <addresses.txt> [output.txt]")
        print("  addresses.txt: file with one hex address per line")
        print("  output.txt: output file for decompiled C (default: decompiled_output.txt)")
        return

    addr_file = args[0]
    out_file = args[1] if len(args) > 1 else os.path.join(os.path.dirname(addr_file), "decompiled_output.txt")

    print("=== Ghidra Batch Decompiler ===")
    print("Binary: " + currentProgram.getName())
    print("Address file: " + addr_file)
    print("Output file: " + out_file)

    # Read target addresses
    addresses = read_addresses(addr_file)
    print("Loaded %d addresses to decompile" % len(addresses))

    # Set up decompiler
    decomp = DecompInterface()
    opts = DecompileOptions()
    decomp.setOptions(opts)
    decomp.openProgram(currentProgram)

    start_time = time.time()
    success = 0
    failed = 0
    skipped = 0

    with open(out_file, 'w') as out:
        out.write("// ============================================================\n")
        out.write("// Black & White Batch Decompilation Output\n")
        out.write("// Binary: %s\n" % currentProgram.getName())
        out.write("// Addresses: %d\n" % len(addresses))
        out.write("// ============================================================\n\n")

        for i, addr_long in enumerate(addresses):
            addr_hex = "0x%08x" % addr_long
            func = get_function_at(currentProgram, addr_long)

            if not func:
                print("[%d/%d] SKIP: %s — no function found" % (i+1, len(addresses), addr_hex))
                out.write("// SKIP: %s — no function at this address\n\n" % addr_hex)
                skipped += 1
                continue

            func_name = func.getName()
            func_size = func.getBody().getNumAddresses()

            try:
                results = decomp.decompileFunction(func, 60, ConsoleTaskMonitor())

                if results.decompileCompleted():
                    c_code = results.getDecompiledFunction().getC()

                    out.write("// ========================================\n")
                    out.write("// [%d] %s\n" % (i+1, func_name))
                    out.write("// Address: %s\n" % addr_hex)
                    out.write("// Size: %d bytes\n" % func_size)
                    out.write("// ========================================\n\n")
                    out.write(c_code)
                    out.write("\n\n")

                    success += 1
                    print("[%d/%d] OK: %s @ %s (%d bytes)" % (i+1, len(addresses), func_name, addr_hex, func_size))
                else:
                    error_msg = results.getErrorMessage() if results.getErrorMessage() else "unknown error"
                    out.write("// FAILED: %s @ %s — %s\n\n" % (func_name, addr_hex, error_msg))
                    failed += 1
                    print("[%d/%d] FAILED: %s @ %s — %s" % (i+1, len(addresses), func_name, addr_hex, error_msg))

            except Exception as e:
                out.write("// ERROR: %s @ %s — %s\n\n" % (func_name, addr_hex, str(e)))
                failed += 1
                print("[%d/%d] ERROR: %s — %s" % (i+1, len(addresses), func_name, str(e)))

        elapsed = time.time() - start_time

        out.write("// ============================================================\n")
        out.write("// SUMMARY\n")
        out.write("// Total: %d | Success: %d | Failed: %d | Skipped: %d\n" % (len(addresses), success, failed, skipped))
        out.write("// Time: %.1f seconds (%.2f func/sec)\n" % (elapsed, success / max(elapsed, 0.001)))
        out.write("// ============================================================\n")

    decomp.closeProgram()

    print("\n=== Complete ===")
    print("Success: %d / %d" % (success, len(addresses)))
    print("Failed: %d, Skipped: %d" % (failed, skipped))
    print("Time: %.1f seconds" % elapsed)
    print("Output: %s" % out_file)

run()
